src/infra/core/src/query_service_impl.rs

```diff
@@ -74,6 +81,15 @@ impl QueryServiceImpl {
         dataset_ref: &DatasetRef,
     ) -> Result<Option<Type>, QueryError> {
         let dataset_handle = self.dataset_repo.resolve_dataset_ref(dataset_ref).await?;
+
+        self.dataset_action_authorizer
```

Comment:
I wonder on which level(s) we should enforce auth. Our data / metadata access is almost entirely hidden behind `MetadataChain` and `ObjectRepository` interfaces, so perhaps we could enforce rules there (although it will require a good caching not to repeat same checks on every block or file read). Checking on service level is nice and coars-grained, but leaves a significant chance to forget the check.

Reply:
In EE world, security checks like this are done at the Service Layer level, not at the Data level.